---
layout: post
title: "Running VEP with GTF files"
date: 2025-01-17
author: "Amit Indap"
tags:
  - bioinformatics
---

The [Variant Effect Predictor](https://github.com/Ensembl/ensembl-vep) (VEP) is a popular tool for annotating genetic variants and it seems like has been around forever.
Typically, running VEP requires the use of a cache file, which is a downloadable file containing  transcript models and other features requried for variant annotation. 

Well, I was today years old when I realized that [VEP can also be run with GTF files](https://useast.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gff). 
This is a great option for those who want to run VEP without downloading the cache file.

I pulled down Gencode GTF files for human [here](https://downloads.pacbcloud.com/public/dataset/MAS-Seq/REF-pigeon_ref_sets/Human_hg38_Gencode_v39/).

I have a Nextflow process for running VEP with GTF files [here](https://github.com/indapa/nextflow-wave-indapa/blob/master/modules/ensemblvep/main.nf)

But the basic command is:

```
vep -i ${phased_vcf} -o ${sample_id}.deepvariant.phased.vep.vcf.gz --format vcf --gtf ${pigeon_gtf} --fasta ${reference} --vcf --everything --fork 8 --compress_output bgzip
```

Even after using a VEP  for many years, there is always something new to learn!



